packages/cdktf-cli/bin/cmds/ui/watch.tsx

```diff
@@ -0,0 +1,271 @@
+}) => {
+  let color: TextProps["color"] | undefined = undefined;
+  let icon: string | React.ReactNode = " ";
+  switch (resource.deployState) {
+    case "CREATING":
+      color = "greenBright";
+      icon = <Spinner type="dots" />;
+      break;
+    case "CREATED":
+      color = "greenBright";
+      icon = "+";
+      break;
+    case "UPDATING":
+      icon = <Spinner type="dots" />;
+      color = "yellowBright";
+      break;
+    case "UPDATED":
+      color = "yellowBright";
+      icon = "~";
+      break;
+    case "DESTROYING":
+      icon = <Spinner type="dots" />;
+      color = "redBright";
+      break;
+    case "DESTROYED":
+      color = "redBright";
+      icon = "-";
+      break;
+    case "WAITING":
+      icon = <Spinner type="dots" />;
+      break;
+    case "ERROR":
+    case "SUCCESS":
+      break;
+  }
+
+  const [columns] = useStdoutDimensions();
+
+  return (
+    <Box width={columns}>
+      <Text color={color}>
+        {icon} {resource.id}
+      </Text>
+      <Spacer />
+      <Text color="gray">{format(new Date(resource.changedAt), "pp")}</Text>
+    </Box>
+  );
+};
+
+const resourceIsInProgress = (
+  resource: GraphQLWatchState["resources"][number]
+) => {
+  return ["WAITING", "UPDATING", "CREATING", "DESTROYING"].includes(
+    resource.deployState
+  );
+};
+
+const DeployingResources = ({
+  resources,
+}: {
+  resources?: GraphQLWatchState["resources"];
+}) => {
+  const inProgressResources = useMemo(
+    () => resources?.filter(resourceIsInProgress) || [],
+    [resources]
+  );
+
+  return (
+    <Box>
+      {inProgressResources.map((r) => (
+        <DeployedResource key={r.id} resource={r} />
+      ))}
+    </Box>
+  );
+};
+
+const DeployedResources = ({
+  resources,
+}: {
+  resources?: GraphQLWatchState["resources"];
+}) => {
+  const [finishedResourcesHistory, setFinishedResourcesHistory] = useState<
+    GraphQLWatchState["resources"]
+  >([]);
+
+  useEffect(() => {
+    const doneResources =
+      resources?.filter((r) => !resourceIsInProgress(r)) || [];
+    setFinishedResourcesHistory((hist) => {
+      const newDoneResources = doneResources.filter(
+        (newDone) =>
+          // skips items that already exist with the same id and timestamp
+          !hist.some(
+            (existingDone) =>
+              existingDone.id === newDone.id &&
+              existingDone.changedAt === newDone.changedAt
+          )
+      );
+      return hist.concat(newDoneResources);
+    });
+  }, [resources]);
+
+  return (
+    <Static items={finishedResourcesHistory}>
+      {(r) => <DeployedResource key={r.id} resource={r} />}
+    </Static>
+  );
+};
+
+const ErrorComponent = ({
+  error,
+}: {
+  error: NonNullable<GraphQLWatchState["error"]>;
+}) => {
+  if (!error.recoverable) {
+    throw new Error(`${error.origin}: ${error.message}`);
+  }
+  return (
+    <Text color="redBright">
+      {error.origin}: {stripAnsi(error.message)}
+    </Text>
+  );
+};
+
+interface WatchConfig {
```

Comment:
Nitpick: https://www.graphql-code-generator.com/ could help here :)

Reply:
Jup, that's something for the future – I mentioned generating the types somewhere else in some Todo comment.